### PR TITLE
mqtt: raise event on parse error -- v2

### PR DESCRIPTION
**Pull request by satta**
Previous PR: #6988

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5018

Describe changes:
* Unify syntax for `where` statement in nom 5+ functions.
* Add `#[inline]` directives where possible.
* Mark more functions as non-public.

suricata-verify-pr: 737

### CODE REVIEW

**Comment by victorjulien** [comment]
Keep in mind that when specifying the suricata verify pr the `#` sign acts to make the line a comment, so remove it to make it effective. I've updated it now

**Comment by victorjulien** [comment]
Can you also add a ticket reference to the PR?

**Comment by satta** [comment]
> Can you also add a ticket reference to the PR?

It was linked in the previous PR. Will now keep it in all following PRs as well. Added to this PR.

**Comment by victorjulien** [comment]
> > Can you also add a ticket reference to the PR?
> 
> It was linked in the previous PR. Will now keep it in all following PRs as well. Added to this PR.

Working on tooling that checks PRs and a missing ticket will be one of the things that will push it into a slow path :)

**Comment by satta** [comment]
Cool, got it. I was assuming that the ticket was only needed for the initial PR. 👍🏻  Looking forward to the upcoming automation!

File: rust/src/mqtt/parser.rs [hunk 1]
```diff
@@ -222,99 +222,111 @@ pub fn parse_connect(i: &[u8]) -> IResult<&[u8], MQTTConnectData> {
...
-        i,
-        MQTTPublishData {
-            topic,
-            message_id,
-            message: message.to_vec(),
-            properties,
-        },
-    ))
+#[inline]
+fn parse_publish(protocol_version: u8, has_id: bool) -> impl Fn(&[u8]) -> IResult<&[u8], MQTTPublishData>
```
Comment (jasonish) [inline rust/src/mqtt/parser.rs:245]: I get it, but I find that functions that return functions often make things more complicate than they are.  Is there another way?

Comment (satta) [reply rust/src/mqtt/parser.rs:245]: I'm afraid that is the way it's done now if you want to have a parser combinator that takes more parameters than just the input, you need to produce a function that takes just the input. At least that's how I understood it.

File: rust/src/mqtt/parser.rs [hunk 2]
```diff
@@ -222,94 +222,104 @@ pub fn parse_connect(i: &[u8]) -> IResult<&[u8], MQTTConnectData> {
...
-    let (i, properties) = parse_properties(i, protocol_version == 5)?;
-    Ok((
-        i,
-        MQTTConnackData {
-            session_present: (topic_name_compression_response & 1) != 0,
-            return_code,
-            properties,
-        },
-    ))
+pub fn parse_connack(protocol_version: u8) -> impl Fn(&[u8]) -> IResult<&[u8], MQTTConnackData>
```
Comment (catenacyber) [inline rust/src/mqtt/parser.rs]: These formatting changes should go into their own commit, right ?

File: rust/src/mqtt/parser.rs [hunk 3]
```diff
@@ -606,7 +635,7 @@ pub fn parse_message(
                     };
                     return Ok((&input[skiplen + len..], msg));
                 }
-                MQTTTypeCode::DISCONNECT => match parse_disconnect(rem, len, protocol_version) {
+                MQTTTypeCode::DISCONNECT => match complete(parse_disconnect(len, protocol_version))(rem) {
```
Comment (catenacyber) [inline rust/src/mqtt/parser.rs:643]: `parse_auth` is missing its `complete` addition, right ?

File: rust/src/mqtt/parser.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/mqtt/parser.rs:245]: Thinking about this, I bet you can handle the `Incomplete` in only one place if you create the right intermediate function
`parse_message` calls some intermediate function to do `match message_type {`
If this intermediate function returns `Incomplete, `parse_message` transforms it into a plain error

What do you think ?

File: rules/mqtt-events.rules [hunk 4]
```diff
@@ -14,3 +14,4 @@ alert mqtt any any -> any any (msg:"SURICATA MQTT invalid QOS level"; app-layer-
 alert mqtt any any -> any any (msg:"SURICATA MQTT missing message ID"; app-layer-event:mqtt.missing_msg_id; classtype:protocol-command-decode; sid:2229007; rev:1;)
 alert mqtt any any -> any any (msg:"SURICATA MQTT unassigned message type (0 or >15)"; app-layer-event:mqtt.unassigned_msg_type; classtype:protocol-command-decode; sid:2229008; rev:1;)
 alert mqtt any any -> any any (msg:"SURICATA MQTT too many transactions"; app-layer-event:mqtt.too_many_transactions; classtype:protocol-command-decode; sid:2229009; rev:1;)
+alert mqtt any any -> any any (msg:"SURICATA MQTT parse error"; app-layer-event:mqtt.parse_error; classtype:protocol-command-decode; sid:2229010; rev:1;)
```
Comment (catenacyber) [inline rules/mqtt-events.rules:17]: Should be in the previous commit

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the work Sascha
I think it can get simpler with the right reformatting of code

File: rust/src/mqtt/parser.rs (hunk 2 above)
Comment (satta) [reply rust/src/mqtt/parser.rs]: Difficult to separate without undoing a lot of stuff, mixed the `pub` removal with the introduction of the new function structure a bit and then the `rustfmt` also did its thing and changed formatting at the end... Cleaning this up will add quite a bit of manual work to do revert parts of the code to an older state just to get the commits right. Just figured that there was better use of my time... unless you're saying it's a must.

Comment (catenacyber) [reply rust/src/mqtt/parser.rs]: The important comment here is indeed another one than Rustfmt thing :

> Thinking about this, I bet you can handle the Incomplete in only one place if you create the right intermediate function parse_message calls some intermediate function to do match message_type {
If this intermediate function returns Incomplete, parse_message` transforms it into a plain error

What do you think ?

Comment (satta) [reply rust/src/mqtt/parser.rs]: Yes, I was aware of that and am working on a solution :)

File: rust/src/mqtt/parser.rs (hunk 3 above)
Comment (satta) [reply rust/src/mqtt/parser.rs:643]: Yep, done in next commit. Thanks for pointing it out.

File: rust/src/mqtt/parser.rs (hunk 1 above)
Comment (satta) [reply rust/src/mqtt/parser.rs:245]: True. See next PR.

**Comment by satta** [comment]
Next PR: #7129

File: rust/src/mqtt/parser.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/mqtt/parser.rs:245]: Looks good in next PR
